pyquil/quilbase.py

```diff
@@ -189,104 +194,63 @@ def __init__(
         qubits: Iterable[Union[Qubit, QubitPlaceholder, FormalArgument]],
     ):
-        if not isinstance(name, str):
-            raise TypeError("Gate name must be a string")
-
-        if name in RESERVED_WORDS:
-            raise ValueError("Cannot use {} for a gate name since it's a reserved word".format(name))
-
-        if not isinstance(params, collections.abc.Iterable):
-            raise TypeError("Gate params must be an Iterable")
-
-        if not isinstance(qubits, collections.abc.Iterable):
-            raise TypeError("Gate arguments must be an Iterable")
-
-        for qubit in qubits:
-            if not isinstance(qubit, (Qubit, QubitPlaceholder, FormalArgument)):
-                raise TypeError("Gate arguments must all be Qubits")
-
-        qubits_list = list(qubits)
-        if len(qubits_list) == 0:
-            raise TypeError("Gate arguments must be non-empty")
-
-        self.name = name
-        self.params = list(params)
-        self.qubits = qubits_list
-        self.modifiers: List[str] = []
+        self._gate = RSGate(name, convert_to_rs_expressions(params), convert_to_rs_qubits(qubits), [])
 
     def get_qubits(self, indices: bool = True) -> Set[QubitDesignator]:
-        return {_extract_qubit_index(q, indices) for q in self.qubits}
-
-    def out(self) -> str:
-        if self.params:
-            return "{}{}{} {}".format(
-                " ".join(self.modifiers) + " " if self.modifiers else "",
-                self.name,
-                _format_params(self.params),
-                _format_qubits_out(self.qubits),
-            )
+        """
+        .. deprecated:: 4.0
+            The indices flag will be separated, use get_qubit_indices() instead
+        """
+        if indices:
+            return self.get_qubit_indices()
         else:
-            return "{}{} {}".format(
-                " ".join(self.modifiers) + " " if self.modifiers else "",
-                self.name,
-                _format_qubits_out(self.qubits),
-            )
+            return set(self._gate.qubits)
+
+    def get_qubit_indices(self) -> Set[int]:
+        return {qubit.as_fixed() for qubit in self._gate.qubits}
 
     def controlled(self, control_qubit: Union[QubitDesignator, Sequence[QubitDesignator]]) -> "Gate":
         """
         Add the CONTROLLED modifier to the gate with the given control qubit or Sequence of control
         qubits.
         """
-        control_qubit = control_qubit if isinstance(control_qubit, Sequence) else [control_qubit]
-        for qubit in control_qubit:
-            qubit = unpack_qubit(qubit)
-            self.modifiers.insert(0, "CONTROLLED")
-            self.qubits.insert(0, qubit)
+        if isinstance(control_qubit, Sequence):
+            for qubit in control_qubit:
+                self._gate = self._gate.controlled(convert_to_rs_qubit(qubit))
+        else:
+            self._gate = self._gate.controlled(convert_to_rs_qubit(control_qubit))
```

Comment:
Oh, good question. I'd say that might cause all kinds of ugly breaks in user scripts. Is is just about closely tracking the python bindings underneath or is there another need to make that change?

Reply:
Yeah, that and I figure we'll want to move away from this pattern eventually. That said, it would be a really ugly break, so maybe it's best left to the future.